File: leetcode/LinkedList/python/21 - Merge two sorted lists.py
# https://leetcode.com/submissions/detail/115678607/

'''
Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes of the first two lists.
'''




# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        a, b = l1, l2
        head = ListNode(-1)
        curr = head
        while a and b:
            #put the bigger on 'a'
            if a.val > b.val:
                a, b = b, a            
            #just travel or point to the bigger
            curr.next = a 
            a = a.next            
            curr = curr.next
        curr.next = a or b
        
        return head.next

File: leetcode/LinkedList/java/21 - Merge two sorted lists.java
//https://leetcode.com/submissions/detail/115713169/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */


class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode a = l1;
        ListNode b = l2;
        ListNode head = new ListNode(-1);
        ListNode curr = head;
        while (a != null && b != null){
            if (a.val > b.val){
                ListNode tmp = a;
                a = b;
                b = tmp;
            }
            curr.next = a;
            a = a.next;
            curr = curr.next;            
        }
        
        if (a != null){
            curr.next = a;
        }
        
        if (b != null){
            curr.next = b;
        }
        return head.next;
    }
}